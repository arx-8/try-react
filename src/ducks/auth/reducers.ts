import { AuthToken } from "domain/models/Auth"
import produce from "immer"
import { reducerWithInitialState } from "typescript-fsa-reducers"
import { actions } from "./actions"

export type State = Readonly<{
  authToken?: AuthToken
}>

export const initialState: State = {
  authToken: undefined,
}

export const reducer = reducerWithInitialState(initialState)
  .case(actions.login, (state) => {
    return produce(state, (draft) => {
      draft.authToken = "any auth token" as AuthToken
    })
  })
  .case(actions.logout, (state) => {
    return produce(state, (draft) => {
      draft.authToken = undefined
    })
  })
  .build()
