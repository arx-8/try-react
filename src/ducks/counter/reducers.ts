import * as actions from "./actions"
import { CounterActionType } from "./types"
import { handleActions, ReduxCompatibleReducer } from "redux-actions"
import produce from "immer"

export type CounterState = Readonly<{
  count: number
  defaultAmount: number
}>

// redux-actions の型がうまく書けない。redux-actions はだいぶダメな感じ。
type Action =
  | ReturnType<
      | typeof actions.add
      | typeof actions.changeDefaultAmount
      | typeof actions.decrement
      | typeof actions.increment
    >
  | any

export const initialState: CounterState = {
  count: 0,
  defaultAmount: 1,
}

export const reducer: ReduxCompatibleReducer<
  CounterState,
  Action
> = handleActions(
  {
    [CounterActionType.ADD]: (state, action) => {
      return produce(state, (draft) => {
        draft.count = state.count + action.payload.count
      })
    },
    [CounterActionType.DECREMENT]: (state) => {
      return produce(state, (draft) => {
        draft.count = state.count - 1
      })
    },
    [CounterActionType.INCREMENT]: (state) => {
      return produce(state, (draft) => {
        draft.count = state.count + 1
      })
    },
    [CounterActionType.CHANGE_DEFAULT_AMOUNT]: (state, action) => {
      return produce(state, (draft) => {
        draft.defaultAmount = action.payload.defaultAmount
      })
    },
  },
  initialState
)
