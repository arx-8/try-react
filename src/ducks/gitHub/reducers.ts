import { Action, ActionTypes } from "./actions"
import { GitHubState } from "./types"
import { Reducer } from "redux"
import produce from "immer"

export const initialState: GitHubState = {
  users: [],
  isLoading: false,
  error: undefined,
}

export const reducer: Reducer<GitHubState, Action> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case ActionTypes.FETCH_MEMBERS_START:
      return produce(state, (draft) => {
        draft.isLoading = true
      })

    case ActionTypes.FETCH_MEMBERS_SUCCEED:
      return produce(state, (draft) => {
        const { users } = action.payload
        draft.users = users
        draft.isLoading = false
      })

    case ActionTypes.FETCH_MEMBERS_FAIL:
      return produce(state, (draft) => {
        const { error } = action.payload
        draft.error = error
        draft.isLoading = false
      })

    case ActionTypes.INITIALIZE_MEMBERS:
      return produce(state, (draft) => {
        draft.users = []
      })

    default: {
      // case の定義忘れ防止のため
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const _: never = action
      return state
    }
  }
}
