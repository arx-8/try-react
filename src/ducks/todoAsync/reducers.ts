import { Todo, VisibilityFilter, TodoId } from "domain/models/Todo"
import produce from "immer"
import { reducerWithInitialState } from "typescript-fsa-reducers"
import { actions } from "./actions"

export type State = Readonly<{
  todoList: Todo[]
  visibilityFilter: VisibilityFilter
  loading: {
    all: boolean
    add: boolean
    ids: TodoId[]
  }
  errorMessage?: string
}>

export const initialState: State = {
  todoList: [],
  visibilityFilter: "all",
  loading: {
    all: false,
    add: false,
    ids: [],
  },
  errorMessage: undefined,
}

export const reducer = reducerWithInitialState(initialState)
  /**
   * shared
   */
  .cases(
    [actions.changeTodoStatus.done, actions.deleteTodo.done],
    (state, payload) => {
      return produce(state, (draft) => {
        draft.loading.ids = draft.loading.ids.filter(
          (id) => id !== payload.params.id
        )
        draft.errorMessage = undefined
      })
    }
  )
  .cases(
    [actions.changeTodoStatus.failed, actions.deleteTodo.failed],
    (state, payload) => {
      return produce(state, (draft) => {
        draft.loading.ids = draft.loading.ids.filter(
          (id) => id !== payload.params.id
        )
        draft.errorMessage = payload.error.message
      })
    }
  )
  /**
   * changeTodoStatus
   */
  .case(actions.changeTodoStatus.started, (state, payload) => {
    return produce(state, (draft) => {
      const { id, label, status } = payload
      draft.loading.ids.push(id)

      // Optimistic Updates
      const t = draft.todoList.find((t) => t.id === id)!
      if (label) {
        t.label = label
      }
      if (status) {
        t.status = status
      }
    })
  })
  /**
   * deleteTodo
   */
  .case(actions.deleteTodo.started, (state, payload) => {
    return produce(state, (draft) => {
      const { id } = payload
      draft.loading.ids.push(id)

      // Optimistic Updates
      draft.todoList = draft.todoList.filter((t) => t.id !== id)!
    })
  })
  /**
   * addTodo
   */
  .case(actions.addTodo.started, (state) => {
    return produce(state, (draft) => {
      draft.loading.add = true
    })
  })
  .case(actions.addTodo.done, (state) => {
    return produce(state, (draft) => {
      draft.loading.add = false
      draft.errorMessage = undefined
    })
  })
  .case(actions.addTodo.failed, (state, payload) => {
    return produce(state, (draft) => {
      draft.loading.add = false
      draft.errorMessage = payload.error.message
    })
  })
  /**
   * fetchAllTodos
   */
  .case(actions.fetchAllTodos.started, (state) => {
    return produce(state, (draft) => {
      draft.loading.all = true
    })
  })
  .case(actions.fetchAllTodos.done, (state, payload) => {
    return produce(state, (draft) => {
      const { result } = payload
      draft.todoList = result
      draft.loading.all = false
      draft.errorMessage = undefined
    })
  })
  .case(actions.fetchAllTodos.failed, (state, payload) => {
    return produce(state, (draft) => {
      draft.loading.all = false
      draft.errorMessage = payload.error.message
    })
  })
  /**
   * setVisibilityFilter
   */
  .case(actions.setVisibilityFilter, (state, payload) => {
    return produce(state, (draft) => {
      const { visibilityFilter } = payload
      draft.visibilityFilter = visibilityFilter
    })
  })
  .build()
