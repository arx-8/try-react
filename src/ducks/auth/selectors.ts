import { State } from "./reducers"

export const isAuthenticated = (state: State): boolean => {
  return !!state.authToken
}
