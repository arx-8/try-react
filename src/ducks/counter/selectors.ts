import { CounterState } from "./reducers"

const get10PerCount = (counter: CounterState): number => {
  return Math.ceil(counter.count / 10)
}

export const selectors = {
  get10PerCount,
}
