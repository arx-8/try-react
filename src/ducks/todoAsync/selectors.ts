import { Todo } from "domain/models/Todo"
import { State } from "./reducers"

const filterTodoList = (state: State): Todo[] => {
  if (state.visibilityFilter === "all") {
    return state.todoList
  }
  return state.todoList.filter((t) => t.status === state.visibilityFilter)
}

const isSomeLoading = (state: State): boolean => {
  return state.loading.all || 0 < state.loading.ids.length
}

export const selectors = {
  filterTodoList,
  isSomeLoading,
}
