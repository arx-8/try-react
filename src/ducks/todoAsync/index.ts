import { actions } from "./actions"
import { operations } from "./operations"
import { reducer, State } from "./reducers"
import { selectors } from "./selectors"

export const todoAsyncActions = actions
export const todoAsyncOperations = operations
export const todoAsyncReducer = reducer
export const todoAsyncSelectors = selectors
export type TodoAsyncState = State
