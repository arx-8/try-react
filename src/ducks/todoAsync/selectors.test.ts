import { TodoId, VisibilityFilter } from "domain/models/Todo"
import { RootState } from "ducks/store"
import produce from "immer"
import { initialState } from "./reducers"
import { selectors } from "./selectors"

/**
 * テストに必要な最小限の RootState を生成して返す
 */
const createInitialRootState = (): RootState => {
  return {
    todoAsync: initialState,
  } as any
}

describe("filterTodoList", () => {
  it("by initialState", () => {
    expect.hasAssertions()
    // ## Arrange ##
    const currentState = createInitialRootState()
    // ## Act ##
    const result = selectors.filterTodoList(currentState)
    // ## Assert ##
    expect(result).toMatchSnapshot()
  })

  it("todoList 1件以上, filter:all", () => {
    expect.hasAssertions()
    // ## Arrange ##
    const currentState = produce(createInitialRootState(), (draft) => {
      draft.todoAsync.todoList.push({
        id: "id1" as TodoId,
        label: "id1-label",
        status: "active",
      })
      draft.todoAsync.todoList.push({
        id: "id2" as TodoId,
        label: "id2-label",
        status: "completed",
      })
      draft.todoAsync.visibilityFilter = "all" as VisibilityFilter
    })
    // ## Act ##
    const result = selectors.filterTodoList(currentState)
    // ## Assert ##
    expect(result).toMatchSnapshot()
  })

  it("todoList 1件以上, filter:completed", () => {
    expect.hasAssertions()
    // ## Arrange ##
    const currentState = produce(createInitialRootState(), (draft) => {
      draft.todoAsync.todoList.push({
        id: "id1" as TodoId,
        label: "id1-label",
        status: "active",
      })
      draft.todoAsync.todoList.push({
        id: "id2" as TodoId,
        label: "id2-label",
        status: "completed",
      })
      draft.todoAsync.todoList.push({
        id: "id3" as TodoId,
        label: "id3-label",
        status: "completed",
      })
      draft.todoAsync.todoList.push({
        id: "id4" as TodoId,
        label: "id4-label",
        status: "completed",
      })
      draft.todoAsync.visibilityFilter = "completed" as VisibilityFilter
    })
    // ## Act ##
    const result = selectors.filterTodoList(currentState)
    // ## Assert ##
    expect(result).toMatchSnapshot()
  })
})
