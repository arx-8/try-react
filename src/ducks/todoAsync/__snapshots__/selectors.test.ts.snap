// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterTodoList by initialState 1`] = `Array []`;

exports[`filterTodoList todoList 1件以上, filter:all 1`] = `
Array [
  Object {
    "id": "id1",
    "label": "id1-label",
    "status": "active",
  },
  Object {
    "id": "id2",
    "label": "id2-label",
    "status": "completed",
  },
]
`;

exports[`filterTodoList todoList 1件以上, filter:completed 1`] = `
Array [
  Object {
    "id": "id2",
    "label": "id2-label",
    "status": "completed",
  },
  Object {
    "id": "id3",
    "label": "id3-label",
    "status": "completed",
  },
  Object {
    "id": "id4",
    "label": "id4-label",
    "status": "completed",
  },
]
`;
